autosummary crashes with wrong autosummary_generate

**Describe the bug**
Sphinx and autosummary are crashed if autosummary_generate has wrong value.

**To Reproduce**
Steps to reproduce the behavior:
```
.. autosummary::
   :toctree: generated

   example.Base
```
```
autosummary_generate = ['foo']
extensions.append('sphinx.ext.autosummary')
```
```
Running Sphinx v2.2.0+
making output directory... done
[autosummary] generating autosummary for: foo.rst

Exception occurred:
  File "/Users/tkomiya/work/sphinx/sphinx/ext/autosummary/generate.py", line 279, in find_autosummary_in_files
    with open(filename, encoding='utf-8', errors='ignore') as f:
FileNotFoundError: [Errno 2] No such file or directory: '/Users/tkomiya/work/tmp/doc/foo.rst'
The full traceback has been saved in /var/folders/rj/80_5mv452nsfnhkvkyqjwj040000gp/T/sphinx-err-je2ykn1a.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
make: *** [html] Error 2
```
```
# Loaded extensions:
Traceback (most recent call last):
  File "/Users/tkomiya/work/sphinx/sphinx/cmd/build.py", line 283, in build_main
    args.tags, args.verbosity, args.jobs, args.keep_going)
  File "/Users/tkomiya/work/sphinx/sphinx/application.py", line 268, in __init__
    self._init_builder()
  File "/Users/tkomiya/work/sphinx/sphinx/application.py", line 329, in _init_builder
    self.events.emit('builder-inited')
  File "/Users/tkomiya/work/sphinx/sphinx/events.py", line 103, in emit
    results.append(callback(self.app, *args))
  File "/Users/tkomiya/work/sphinx/sphinx/ext/autosummary/__init__.py", line 758, in process_generate_options
    app=app, imported_members=imported_members)
  File "/Users/tkomiya/work/sphinx/sphinx/ext/autosummary/generate.py", line 168, in generate_autosummary_docs
    items = find_autosummary_in_files(sources)
  File "/Users/tkomiya/work/sphinx/sphinx/ext/autosummary/generate.py", line 279, in find_autosummary_in_files
    with open(filename, encoding='utf-8', errors='ignore') as f:
FileNotFoundError: [Errno 2] No such file or directory: '/Users/tkomiya/work/tmp/doc/foo.rst'
```

**Expected behavior**
Build successfully (with warning)

**Your project**
None

**Screenshots**
None

**Environment info**
- OS: Mac
- Python version: 3.7.3
- Sphinx version: 2.1.1
- Sphinx extensions:  sphinx.ext.autosummary

--- a/CHANGES
+++ b/CHANGES
@@ -41,6 +41,7 @@ Bugs fixed
 * #5502: linkcheck: Consider HTTP 503 response as not an error
 * #6439: Make generated download links reproducible
 * #6486: UnboundLocalError is raised if broken extension installed
+* #6498: autosummary: crashed with wrong autosummary_generate setting
 
 Testing
 --------

--- a/sphinx/ext/autosummary/__init__.py
+++ b/sphinx/ext/autosummary/__init__.py
@@ -58,6 +58,7 @@ import posixpath
 import re
 import sys
 import warnings
+from os import path
 from types import ModuleType
 from typing import List, cast
 
@@ -731,25 +732,30 @@ def process_generate_options(app):
     # type: (Sphinx) -> None
     genfiles = app.config.autosummary_generate
 
-    if genfiles and not hasattr(genfiles, '__len__'):
+    if genfiles is True:
         env = app.builder.env
         genfiles = [env.doc2path(x, base=None) for x in env.found_docs
                     if os.path.isfile(env.doc2path(x))]
+    else:
+        ext = list(app.config.source_suffix)
+        genfiles = [genfile + (not genfile.endswith(tuple(ext)) and ext[0] or '')
+                    for genfile in genfiles]
+
+        for entry in genfiles[:]:
+            if not path.isfile(path.join(app.srcdir, entry)):
+                logger.warning(__('autosummary_generate: file not found: %s'), entry)
+                genfiles.remove(entry)
 
     if not genfiles:
         return
-
-    from sphinx.ext.autosummary.generate import generate_autosummary_docs
-
-    ext = list(app.config.source_suffix)
-    genfiles = [genfile + (not genfile.endswith(tuple(ext)) and ext[0] or '')
-                for genfile in genfiles]
 
     suffix = get_rst_suffix(app)
     if suffix is None:
         logger.warning(__('autosummary generats .rst files internally. '
                           'But your source_suffix does not contain .rst. Skipped.'))
         return
+
+    from sphinx.ext.autosummary.generate import generate_autosummary_docs
 
     imported_members = app.config.autosummary_imported_members
     with mock(app.config.autosummary_mock_imports):

--- a/tests/test_ext_autosummary.py
+++ b/tests/test_ext_autosummary.py
@@ -302,3 +302,9 @@ def test_generate_autosummary_docs_property(app):
                        ".. currentmodule:: target.methods\n"
                        "\n"
                        ".. autoproperty:: Base.prop")
+
+
+@pytest.mark.sphinx('dummy', testroot='ext-autosummary',
+                    confoverrides={'autosummary_generate': ['unknown']})
+def test_invalid_autosummary_generate(app, status, warning):
+    assert 'WARNING: autosummary_generate: file not found: unknown.rst' in warning.getvalue()

